internal/transport/grpcframer/framer.go

```diff
@@ -0,0 +1,267 @@
+type Flags uint8
+
+const (
+	// Data Frame
+	FlagDataEndStream Flags = 0x1
+	FlagDataPadded    Flags = 0x8
+
+	// Headers Frame
+	FlagHeadersEndStream  Flags = 0x1
+	FlagHeadersEndHeaders Flags = 0x4
+	FlagHeadersPadded     Flags = 0x8
+	FlagHeadersPriority   Flags = 0x20
+
+	// Settings Frame
+	FlagSettingsAck Flags = 0x1
+
+	// Ping Frame
+	FlagPingAck Flags = 0x1
+
+	// Continuation Frame
+	FlagContinuationEndHeaders Flags = 0x4
+
+	FlagPushPromiseEndHeaders Flags = 0x4
+	FlagPushPromisePadded     Flags = 0x8
+)
+
+// Setting represents the id and value pair of a particular HTTP/2 setting.
+// See https://httpwg.org/specs/rfc7540.html#SettingFormat.
+type Setting struct {
+	ID    SettingID
+	Value uint32
+}
+
+// SettingID represents the id of a specific HTTP/2 setting.
+// See https://httpwg.org/specs/rfc7540.html#SettingValues.
+type SettingID uint16
+
+const (
+	SettingsHeaderTableSize      SettingID = 0x1
+	SettingsEnablePush           SettingID = 0x2
+	SettingsMaxConcurrentStreams SettingID = 0x3
+	SettingsInitialWindowSize    SettingID = 0x4
+	SettingsMaxFrameSize         SettingID = 0x5
+	SettingsMaxHeaderListSize    SettingID = 0x6
+)
+
+// FrameHeader is the 9 byte header of any HTTP/2 Frame.
+// See https://httpwg.org/specs/rfc7540.html#FrameHeader.
+type FrameHeader struct {
+	// Size is the size of the frame's payload without the 9 header bytes.
+	// As per the HTTP/2 spec, size can be up to 3 bytes, but only frames
+	// up to 16KB can be processed without agreement.
+	Size uint32
+
+	// Type is a byte that represents the Frame Type. The HTTP/2 spec
+	// defines 10 standard types but extension frames may be written.
+	Type FrameType
+
+	// Flags is a byte representing the flags set for the specific frame
+	// type.
+	Flags Flags
+
+	// StreamID is the ID for the stream which this frame is for. If the
+	// frame is connection specific instead of stream specific, the
+	// streamID is 0.
+	StreamID uint32
+}
+
+// Frame represents any kind of HTTP/2 Frame, which can be casted into
+// its corresponding type using the Header method provided.
+type Frame interface {
+	// Header returns this frame's Header.
+	Header() FrameHeader
+}
+
+type DataFrame struct {
+	hdr  FrameHeader
+	pool *sync.Pool
+	Data []byte
+}
+
+func (f *DataFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+func (f *DataFrame) Free() {
+	if f.Data == nil {
+		return
+	}
+
+	f.pool.Put(&f.Data)
+	f.Data = nil
+}
+
+type HeadersFrame struct {
+	hdr      FrameHeader
+	pool     *sync.Pool
+	HdrBlock []byte
+}
+
+func (f *HeadersFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+func (f *HeadersFrame) Free() {
+	if f.HdrBlock == nil {
+		return
+	}
+
+	f.pool.Put(&f.HdrBlock)
+	f.HdrBlock = nil
+}
+
+type RSTStreamFrame struct {
+	hdr  FrameHeader
+	Code ErrorCode
+}
+
+func (f *RSTStreamFrame) Header() FrameHeader {
+	return f.hdr
+}
+
+type SettingsFrame struct {
+	hdr      FrameHeader
+	pool     *sync.Pool
+	settings []byte
```

Comment:
In the different exported frame types, I see Exported fields and unexported fields. Why are some fields exported and some unexported? Should all fields be unexported?

Reply:
Let's implement things when we actually start requiring them. Thanks.